web/packages/teleterm/src/ui/services/keyboardShortcuts/getKeyCode.ts

```diff
@@ -0,0 +1,61 @@
+/**
+ * Copyright 2023 Gravitational, Inc
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Based on tabby https://github.com/Eugeny/tabby/blob/7a8108b20d9cbaab636c932ceaf4bacc710d6a40/tabby-core/src/services/hotkeys.util.ts
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2017 Eugene Pankov
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ */
+
+const REGEX_LATIN_KEYNAME = /^[A-Za-z]$/;
+
+export function getKeyCode(event: KeyboardEvent): string {
```

Comment:
> [...] I assume it's to convert browser's KeyboardEvent code to something supported by Electron? If so, the name of the function should reflect that.

No, it has nothing to do with Electron, we use pure browser keyboard events. I will think about a better name.

>Personally I had hard time understanding what's going on until I looked at MDN for KeyboardEvent docs and realized that event.code is just the character when something like Shift+Ctrl+A is pressed.

Actually, the opposite - `code` is a physical key, not the character, see below.  

The function name definitely can be improved, but I realized that this function has a more important problem - it poorly supports layouts other than US QWERTY. Below is my research, this topic is quite complex :)

### Introduction
An event generated by the browser has two properties interesting for us: `key` and `code`. `key` is a generated character (like `@` when you press `Shift+2` on US QWERTY) and `code` is a name of a physical key that was pressed (like `Digit2`).

For example,  when you press `Option+1`, you get the following values in a keyboard event:
```ts
...
altKey: true,
ctrlKey: false,
shiftKey: false,
metaKey: false,
code: "Digit1",
key: "Ń"
```

Below I compared a few ways of interpreting pressed keys.
### By using a `code`
In this case, the pressed keys are independent of the keyboard layout. For example, after you switch to QWERTZ, you still have to press "Command+ physical Z (which is now a logical Y)" for "Command+Z" shortcut.
I tried to implement it in our codebase in the way tabby does, so for A-Z characters we relayed on a reported character (`key`). It fixed the above example, but what about the other keys? Keyboard layouts affect more than just the A-Z keys.

### By using a `key`
Our previous approach (I wanted to change it to allow users creating shortcuts like "Option+A" - relying on `key` here doesn't work).
After I realized that using key codes has some drawbacks I wanted to go back to using a `key`.
I took a look at some posts and often it was a recommend solution https://kevinsimper.medium.com/why-keyboard-shortcuts-and-accessibility-in-other-keyboard-languages-rarely-works-8638abc15e71.
But then I started playing with different layouts and found a problem ⬇️
Take a look at the Czech layout:
<img width="611" alt="image" src="https://user-images.githubusercontent.com/20583051/221600691-a36ca278-4225-48b5-af9e-10a68a9e45fb.png">
Our default shortcut for Tab 1 is "Command+1". On Czech layout there is no "1" on the standard row, so the user has to press "Command+Shift+1". Unfortunately, this additional "Shift" press (required by the character) prevents a correct match. There is no way to detect if some modifier was used to generate a character or is just part of the shortcut (like in "Ctrl+Shift+Tab"). This is common for a lot of layouts that the user has to press a modifier to generate a character that is available on US QWERTY with any modifier.

From what I understand, when you develop an app in Swift, you [would define a shortcut using this approach.](https://developer.apple.com/documentation/swiftui/keyequivalent) It automatically generates a shortcut for a given "character" + modifiers, we are not able to do it.

### By using a `code` with `KeyboardLayoutMap`
Here you read the key `code` and then look for match in [Keyboard Layout Map](https://developer.mozilla.org/en-US/docs/Web/API/KeyboardLayoutMap). It allows for operating on "physical characters" (I didn't find a better name) on a given keyboard layout. 
Here is what the map looks like:
<img width="382" alt="image" src="https://user-images.githubusercontent.com/20583051/221811438-05b5e0b5-5a0d-4b6b-b706-4c92e2623529.png">

On the left there is a physical US QWERTY key, on the right a "physical character" on the active keyboard layout. 

But as you can see, it still doesn't solve the issue we had with "1" on the Czech layout, because there is no such key on this layout!

My suggestion to solve this is to go back to the physical key if the "physical character" is something that does not exist on US QWERTY. In our example the "physical character" is "+". There is no such key on a standard row in US QWERTY, so we should use "Key1" (without the prefix).

It seems to me that this also a [default behaviour for macOS](https://developer.apple.com/documentation/SwiftUI/KeyboardShortcut/Localization-swift.struct#overview). I set an accelerator "Command+1" for a menu item:
US QWERTY
<img width="382" alt="image" src="https://user-images.githubusercontent.com/20583051/221814551-1e7c237c-8c96-4b1b-bf8c-770c3957b881.png">

Czech QWERTY
<img width="382" alt="image" src="https://user-images.githubusercontent.com/20583051/221814664-69a839c2-02cf-4819-858e-abc8af16e8b1.png">

As you can see the the shortcut changed to "Command+" automatically.  

Now, let's try a key that exists on both layouts (but it is in different places):
US QWERTY
<img width="382" alt="image" src="https://user-images.githubusercontent.com/20583051/221824430-16db311d-0db3-4ad6-9232-832814a26e60.png">

Czech QWERTY
<img width="382" alt="image" src="https://user-images.githubusercontent.com/20583051/221824349-0475ba53-ca2f-4c62-b064-f0e06ba291cd.png">

We would have to do the same in `useKeyboardShortcutFormatters` that is used to generate tooltips. A theoretical shortcut "Command+[" it should be displayed as "Command+[" on US QWERTY, but as "Command+ú" on Czech QWERTY. 

It seems to me that we should go with this approach, let me know if you have any thoughts.
Sorry for a long post, let me know if something is unclear.

Reply:
MDN says Keyboard.getLayoutMap is available only in secure contexts so we'd need to pull a dep for this instead.

https://twitter.com/buildsghost/status/1620806694925402113

Reply:
But overall I like the idea and I think we should go for it if it's not overly time consuming to implement.

Reply:
> As I understand this, these are actually two different problems: interpreting which key was pressed and showing a correct accelerator depending on the keyboard layout? Because in the end the shortcuts need to be provided in a single layout map, right?

Technically yes, and we should able to address both using Keyboard.getLayoutMap. To provide a tooltip the for "Command+[" we have to find "[" in the map (but first we need to remove all these 'Key', 'Digit' prefixes) and read the "character" that is assigned to it.

>Because in the end the shortcuts need to be provided in a single layout map, right?

By "single layout map" you mean the active layout? If so, then yes :)

>MDN says Keyboard.getLayoutMap is available only in secure contexts so we'd need to pull a dep for this instead.

Yes it is available only in a secure context, but I tested it in a packaged app and it worked. It looks like Electron provides secure context even when we load `index.html` from a file.

>But overall I like the idea and I think we should go for it if it's not overly time consuming to implement.

I think the most consuming part was the research which is done; probably we won't be ready with the app config for 12.1 anyway, so IMO we can spend a few hours on it. 

I will revert the changes related to `getKeyCode` in this PR and create another.

Reply:
> > Because in the end the shortcuts need to be provided in a single layout map, right?
> 
> By "single layout map" you mean the active layout? If so, then yes :)

I'm thinking about it in the context of the example you brought up of macOS changing the displayed accelerator when the layout changes.

When we read the config we need to assume that the shortcuts are provided in a specific layout. The layout can change at any time and so this should be reflected by the UI.

Quoting [that SO answer you linked](https://stackoverflow.com/questions/70211837/keyboard-shortcuts-commands-on-non-latin-alphabet-keyboards-javascript):

> It turns out that [most non-Latin keyboards have 2 alphabets printed](https://www.quora.com/What-do-keyboards-look-like-for-countries-that-dont-use-the-Latin-alphabet), Latin and its own - and the user can switch between layouts.

I don't think we need to be perfect on the first attempt here. There's a couple of problems that I see:

1. `Keyboard.getLayoutMap` returns a promise, I believe this is to future-proof the API in case in the future browsers want to introduce some dialog box similar to when a website asks you for your location or to enable notifications.
2. I don't see how we could update the UI when the layout map changes as there's no API which lets you subscribe to the changes.

I was going to say that our best bet for now is to assume that the shortcuts are provided in the keyboard layout that is set when the app starts. But I don't think that's necessarily true? We have default shortcuts that which should stay "logically" consistent, e.g. `Cmd+` on Czech Qwerty should switch to the first tab. I suppose this suggestion of yours is supposed to solve this problem but I'm yet to fully wrap my mind around it:

> My suggestion to solve this is to go back to the physical key if the "physical character" is something that does not exist on US QWERTY. In our example the "physical character" is "+". There is no such key on a standard row in US QWERTY, so we should use "Key1" (without the prefix).

OTOH we don't want to go the other way, that is always assume that the shortcuts are provided in US Qwerty because this would mean that the user would need to provide shortcuts in a way that is totally unnatural for them.

Reply:
>I don't see how we could update the UI when the layout map changes as there's no API which lets you subscribe to the changes.

I was absolutely sure that this API is available https://wicg.github.io/keyboard-map/#layoutchange-event but it's not :(
I wanted to call `Keyboard.getLayoutMap` and cache the response on the layout change.

>OTOH we don't want to go the other way, that is always assume that the shortcuts are provided in US Qwerty because this would mean that the user would need to provide shortcuts in a way that is totally unnatural for them.

That's true, it would be unnatural, but I think that allowing users to provide shortcuts using characters that are not available on US QWERTY would create a mess. Some of the shortcuts would be based on US QWERTY and some on a different layout? What we should do about the latter, convert to US?
Also, we wouldn't be able to validate these characters.

The best solution to solve this is to have a UI, we could just register the keys user pressed and keep them as US QWERTY.

Reply:
>When we read the config we need to assume that the shortcuts are provided in a specific layout. The layout can change at any time and so this should be reflected by the UI.

Could you help me to understand how it would work for a "Command+1" shortcut on a layout where there is no such key?
In my mind, the config should always stay the same.  We would just map the key that was pressed to the key on the active layout.

Reply:
Let's move this conversation to [the Slack thread](https://gravitational.slack.com/archives/C03FJA391M3/p1677680548050679).

Reply:
In the end we decided to stay with Tabby's implementation and interpret only latin keyboard keys by the printed key and the rest by their code.